### PR TITLE
feat: load solc dynamically

### PR DESCRIPTION
**Pull request by o-az**
## Description

_Concise description of proposed changes_

https://github.com/evmts/evmts-monorepo/issues/356 (took a diff path)

## Testing

Since `vitest` doesn't yet support http / network imports, I'm using [node test runner](https://nodejs.org/api/test.html)

## Additional Information

- [ ] I read the [contributing docs](../docs/contributing.md) (if this is your first contribution)

Your ENS/address:

sed.eth

### CODE REVIEW

File: tsconfig/base.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
+	"$schema": "https://json.schemastore.org/tsconfig",
 	"compilerOptions": {
 		"target": "ESNext",
-		"lib": ["esnext"],
+		"lib": ["esnext", "dom", "dom.iterable"],
```
[the remark this answers is not in the document]
Comment (roninjin10) [reply tsconfig/base.json:5]:
```suggestion
		"lib": ["esnext"],
```

File: bundlers/bundler/src/types.ts [hunk 2]
```diff
@@ -118,3 +118,501 @@ export type SolidityResolver = (
 	 */
 	resolveEsmModuleSync: (module: string, basedir: string) => BundlerResult
 }
+
+export interface CompilerOptions<Filename extends string> {
```
Comment (roninjin10) [inline bundlers/bundler/src/types.ts:122]: Where did you get this from? Did you write this yourself?

File: bundlers/bundler/scripts/http-loader.mjs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { get } from 'node:https'
+
+/* https://nodejs.org/api/esm.html#https-loader */
+/**
```
Comment (roninjin10) [inline bundlers/bundler/scripts/http-loader.mjs:4]:
```suggestion
/**
 * @see https://nodejs.org/api/esm.html#https-loader */
```

File: bundlers/bundler/src/compiler/loadSolidityCompilerSync.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * To run:
+ *
+ * ```bash
+ * NODE_OPTIONS='--experimental-loader=./scripts/http-loader.mjs --experimental-network-imports --no-warnings --loader=tsx' node src/compiler/loadSolidityCompilerSync.ts
+ * ```
+ */
+export function getSolidityCompilerSync(): SolidityCompiler {
+	return solc
```
Comment (roninjin10) [inline bundlers/bundler/src/compiler/loadSolidityCompilerSync.ts:13]: this won't work as is because this will have to take a version number.

**Review by roninjin10 — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 5]
```diff
@@ -27,6 +26,7 @@
     "lib": true
   },
   "files.associations": {
-    ".gas-snapshot": "julia"
+    ".gas-snapshot": "julia",
+    "base.json": "jsonc"
```
[the remark this answers is not in the document]
Comment (o-az) [reply .vscode/settings.json:30]: tells vscode that `base.json` is the same file type as `tsconfig.json` (`JSON with Comments`) and that it's ok to have comments in that file (as opposed to `json` file type)

File: bundlers/bundler/src/compiler/loadSolidityCompilerSync.spec.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { getSolidityCompilerSync } from './loadSolidityCompilerSync'
+import * as assert from 'node:assert'
+import { describe, it } from 'node:test'
```
[the remark this answers is not in the document]
Comment (o-az) [reply bundlers/bundler/src/compiler/loadSolidityCompilerSync.spec.ts:3]: can probably get it to work if we switch to `node:vm` or use a intermediary `esbuild` step

relevant 
https://github.com/jestjs/jest/issues/12897

File: bundlers/bundler/src/types.ts (hunk 2 above)
Comment (o-az) [reply bundlers/bundler/src/types.ts:122]: Compiler options: https://docs.soliditylang.org/en/latest/using-the-compiler.html#input-description

Compiler output: https://docs.soliditylang.org/en/latest/metadata.html#contract-metadata

**Comment by roninjin10** [comment]
Added typesafe solc types in this pr https://github.com/evmts/evmts-monorepo/pull/436

**Comment by roninjin10** [comment]
Moving this work to `experimental/solc` folder so we can keep iterating on it without needing to deal with merge conflicts and get this work merged in. Thanks for this!

**Review by roninjin10 — APPROVED** [review]

**Pull request merged**